docs/src/index.md

```diff
@@ -0,0 +1,457 @@
+creation after a "return a table"-style column creation call. 
+
+
+## `@orderby`
+
+Sort rows in a `DataFrame` by values in one of several columns or a 
+transformation of columns. Only operates on `DataFrame`s and not `GroupedDataFrame`s. 
+
+```julia
+df = DataFrame(x = [1, 1, 2, 2], y = [1, 2, 101, 102]);
+@orderby(df, -1 .* :x)
+@orderby(df, :x, :y .- mean(:y))
+```
+
+## `@with`
+
+`@with` creates a scope in which all symbols that appear are aliases for the columns
+in a DataFrame. 
+
+```julia
+df = DataFrame(x = 1:3, y = [2, 1, 2])
+x = [2, 1, 0]
+
+@with(df, :y .+ 1)
+@with(df, :x + x)  # the two x's are different
+
+x = @with df begin
+    res = 0.0
+    for i in 1:length(:x)
+        res += :x[i] * :y[i]
+    end
+    res
+end
+
+@with(df, df[:x .> 1, ^(:y)]) # The ^ means leave the :y alone
+
+```
+
+!!! note 
+    `@with` creates a function, so scope within `@with` is a local scope.
+    Variables in the parent can be read. Writing to variables in the parent scope
+    differs depending on the type of scope of the parent. If the parent scope is a
+    global scope, then a variable cannot be assigned without using the `global` keyword.
+    If the parent scope is a local scope (inside a function or let block for example),
+    the `global` keyword is not needed to assign to that parent scope.
+
+!!! note
+    Because `@with` creates a function, be careful with the use of `return`. 
+
+    ```
+    function data_transform(df; returnearly = false)
+        if returnearly
+            @with df begin 
+                z = :x + :y
+                return z
+            end
+        else 
+            return [1, 2, 3]
+        end
+
+        return [4, 5, 6]
+    end
+    ```
+
+    The above function will return `[4, 5, 6]` because the `return` inside the `@with`
+    applies to the anonymous function created by `@with`. 
+
+    Given that `@eachrow` (below) is implemented with `@with`, the same caveat applies to 
+    `@eachrow` blocks. 
+
+
+## `@eachrow`
+
+Act on each row of a data frame. Includes support for control flow and `begin end` blocks. Since the "environment" induced by `@eachrow df` is implicitly a single row of `df`, one uses regular operators and comparisons instead of their elementwise counterparts as in `@with`. Does not change the input data frame argument.
+
+```julia
+df = DataFrame(A = 1:3, B = [2, 1, 2])
+df2 = @eachrow df begin 
+    :A = :B + 1
+end
+```
+
+`@eachrow` introduces a function scope, so a `let` block is required here to create 
+a scope to allow assignment of variables within `@eachrow`. 
+
+```julia
+df = DataFrame(A = 1:3, B = [2, 1, 2])
+let x = 0.0
+    @eachrow df begin
+        if :A < :B
+            x += :B * :C
+        end
+    end
+    x
+end
+```
+
+`@eachrow` also supports special syntax for allocating new columns to make
+`@eachrow` more useful for data transformations. The syntax `@newcol
+x::Array{Int}` allocates a new column `:x` with an `Array` container with eltype
```

Comment:
This is the first time I notice it. Why `Array{Int}` not `Vector{Int}`? Also can you allocate other types than vectors from Julia Base (e.g. `PooledVector`)?

Reply:
You can allocate any type that would allow this pattern 

```
T(undef, N)
```

So it would probably be pretty hard to construct other types of arrays, I think. 

This is something I would really like to fix but haven't thought of a way to do so without type instabilities.

Reply:
OK - clear. We can leave it for later. Now we document that `Array` is only allowed, so we have flexibility for the future.

Reply:
I have changed all examples to array.